File: services/engine_proxy.py
from core.engine import MacroManager


class EngineProxy:
    def __init__(self, manager: MacroManager):
        self.manager = manager
        self.buttons = {
            **manager.mouse.buttons,
            **manager.keyboard.buttons,
        }

    @property
    def enabled(self):
        return self.manager.mouse.enabled

    @property
    def running(self):
        return self.manager.mouse.running and self.manager.keyboard.running

    @running.setter
    def running(self, state: bool):
        self.manager.mouse.running = state
        self.manager.keyboard.running = state

    def toggle(self):
        self.manager.toggle_all()

    def stop(self):
        self.manager.stop()

    def set_game_safe(self, state: bool):
        self.manager.mouse.game_safe = state
        self.manager.keyboard.game_safe = state

    def set_on_toggle(self, callback):
        self.manager.set_on_toggle(callback)

    def _get_engine(self, key):
        if key in self.manager.mouse.buttons:
            return self.manager.mouse
        return self.manager.keyboard

    def set_cps(self, key, value):
        self._get_engine(key).set_cps(key, value)

    def set_delay(self, key, value):
        self._get_engine(key).set_delay(key, value)

    def set_burst_count(self, key, value):
        self._get_engine(key).set_burst_count(key, value)

    def get_burst_count(self, key):
        return self._get_engine(key).get_burst_count(key)

    def get_cps(self, key):
        return self._get_engine(key).get_cps(key)

    def get_real_cps(self, key):
        return self._get_engine(key).get_real_cps(key)

    def set_active(self, key, state: bool):
        engine = self._get_engine(key)
        if key in engine.buttons:
            engine.buttons[key].active = state

    def is_active(self, key) -> bool:
        engine = self._get_engine(key)
        if key in engine.buttons:
            return engine.buttons[key].active
        return False

File: cpp/src/services/engine_proxy.cpp
#include "services/engine_proxy.hpp"

namespace gx {

EngineProxy::EngineProxy(MacroManager& manager) : manager_(manager) {}

bool EngineProxy::enabled() const { return manager_.mouse().enabled(); }

void EngineProxy::toggle() { manager_.toggle_all(); }

void EngineProxy::stop() { manager_.stop(); }

void EngineProxy::set_game_safe(bool state) {
    manager_.mouse().set_game_safe(state);
    manager_.keyboard().set_game_safe(state);
}

BaseEngine* EngineProxy::engine_for(const std::string& key) {
    if (manager_.mouse().button(key)) return &manager_.mouse();
    return &manager_.keyboard();
}

const BaseEngine* EngineProxy::engine_for(const std::string& key) const {
    if (manager_.mouse().button(key)) return &manager_.mouse();
    return &manager_.keyboard();
}

void EngineProxy::set_cps(const std::string& key, int value) {
    engine_for(key)->set_cps(key, value);
}

void EngineProxy::set_delay(const std::string& key, double value) {
    engine_for(key)->set_delay(key, value);
}

void EngineProxy::set_burst_count(const std::string& key, int value) {
    engine_for(key)->set_burst_count(key, value);
}

int EngineProxy::get_burst_count(const std::string& key) const {
    return engine_for(key)->get_burst_count(key);
}

int EngineProxy::get_cps(const std::string& key) const {
    return engine_for(key)->get_cps(key);
}

int EngineProxy::get_real_cps(const std::string& key) const {
    return engine_for(key)->get_real_cps(key);
}

void EngineProxy::set_active(const std::string& key, bool active) {
    engine_for(key)->set_active(key, active);
}

bool EngineProxy::is_active(const std::string& key) const {
    return engine_for(key)->is_active(key);
}

bool EngineProxy::has_macro(const std::string& key) const {
    return manager_.mouse().button(key) || manager_.keyboard().button(key);
}

int EngineProxy::count_active_macros() const {
    int n = 0;
    for (const auto* eng : {&manager_.mouse(), &manager_.keyboard()}) {
        for (const auto& [k, btn] : eng->buttons()) {
            (void)k;
            if (btn.active) ++n;
        }
    }
    return n;
}

int EngineProxy::get_total_cps() const {
    int total = 0;
    for (const auto* eng : {&manager_.mouse(), &manager_.keyboard()}) {
        for (const auto& [key, btn] : eng->buttons()) {
            if (btn.active) total += eng->get_real_cps(key);
        }
    }
    return total;
}

}  // namespace gx
